File: 1679. Max Number of K-Sum Pairs.py
class Solution:
    # # two pointer on a sorted array
    # # O(n log n) time and O(1) space complexity
    # def maxOperations(self, nums: List[int], k: int) -> int:
    #     nums = sorted(nums)
    #     L, R = 0, len(nums) - 1
    #     out = 0

    #     while L < R:
    #         sum = nums[L] + nums[R]
    #         if sum == k:
    #             out += 1
    #             L += 1
    #             R -= 1
    #         elif sum > k:
    #             R -= 1
    #         else: #sum < k
    #             L += 1

    #     return out

    # hash table solution
    # O(n) space and time complexity
    def maxOperations(self, nums: List[int], k: int) -> int:
        d = defaultdict(int)
        out = 0

        for num in nums:
            d[num] += 1
            comp = k - num

            if num == comp:
                if d[num] > 1:
                    out += 1
                    d[num] -= 2

            elif d[num] > 0 and comp in d and d[comp] > 0:
                out += 1
                d[num] -= 1
                d[comp] -= 1

        return out

File: 1679. Max Number of K-Sum Pairs.ts
// O(n) time and space complexity
function maxOperations(nums: number[], k: number): number {
    let out = 0;
    const d = {};

    nums.forEach((num, i) => {
        if (d[num]) {
            d[num]++;
        } else {
            d[num] = 1;
        };

        let comp = k - num;

        if (comp == num) {
            if (d[num] > 1) {
                out++, d[num] -= 2;
            };
            return
        };
        
        if (d[comp] && d[comp] > 0) {
            out++, d[comp]--, d[num]--;
        };
    });

    return out;
};


// // O(n log n) time and O(n) space complexity
// function maxOperations(nums: number[], k: number): number {
//     const sorted = nums.sort((a, b) => a - b);
//     let L = 0, R = sorted.length - 1;
//     let out = 0;

//     while (L < R) {
//         let sum = sorted[L] + sorted[R];
//         if (sum == k) {
//             out++, L++, R--;
//         } else if (sum > k) {
//             R--;
//         } else {
//             L++;
//         };
//     };

//     return out;
// };
